In [8]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
import statsmodels.formula.api as smf
from patsy import dmatrices

# Overview

<img src="analyze-overview-1.jpg"/>
<br/>
<img src="analyze-overview-2.jpg"/>

Random variable: 
- ước lượng điểm
- khoảng tin cậy

Tiêu chuẩn lựa chọn phương pháp ước lượng

- Unbiased: $\theta = \mathbb{E}(\hat{\theta})$
- Consistent: $\displaystyle \lim_{n \rightarrow \infty} P(|\hat{\theta} - \theta| < \epsilon)$
- Efficient: 

# Ước lượng khoảng tin cậy cho tham số

<img src="517.jpg"/>

In [3]:
sample_y = np.array([1, 3, 7, 9, 5, 2])
sample_x = np.array([(140+145)/2, (145+150)/2, (150+155)/2, (155+160)/2, (160+165)/2, (165+170)/2])

In [4]:
sample_x

array([142.5, 147.5, 152.5, 157.5, 162.5, 167.5])

In [5]:
sample_y

array([1, 3, 7, 9, 5, 2])

In [6]:
sample_x_mean = sample_x @ sample_y.T / np.sum(sample_y)

In [7]:
sample_x_mean

156.2037037037037

In [8]:
sample_x_square_mean = (sample_x - sample_x_mean) ** 2

In [11]:
sample_x_var = sample_x_square_mean @ sample_y.T / (np.sum(sample_y) - 1)

In [12]:
sample_x_var

37.67806267806268

In [13]:
#TODO: b

# Linear Regression Revisited

Estimate confidence interval for $\hat{w}$. Given that $\epsilon = |\hat{y} - y|$, there's an assumption state that $\epsilon_i \sim \mathcal{N}(\mu_i = 0, \sigma_i)$, or, $\sigma = \sigma_i\ \forall i$

Assuming $y \sim [x_1, x_2, .., x_D]$, or, $y = w_0 + w_1x_1 + w_2x_2 + .. + w_Dx_D + \epsilon$

Parameters must be $\neq 0$ to be meaningful

# Hypothesis testing

Quan tâm tới p-value, $p < \alpha$ thì bác bỏ. Làm lại 7.2 dùng statsmodel

In [3]:
df = pd.DataFrame({"income": [31, 50, 47, 45, 39, 50, 35, 40, 45, 50], "expense": [29, 42, 38, 30, 29, 41, 23, 36, 42, 48]})

In [4]:
df

income  expense
0      31       29
1      50       42
2      47       38
3      45       30
4      39       29
5      50       41
6      35       23
7      40       36
8      45       42
9      50       48

In [10]:
results = smf.ols("expense ~ income", data = df).fit()

In [13]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                expense   R-squared:                       0.672
Model:                            OLS   Adj. R-squared:                  0.631
Method:                 Least Squares   F-statistic:                     16.39
Date:                Sat, 20 May 2023   Prob (F-statistic):            0.00369
Time:                        10:48:40   Log-Likelihood:                -28.630
No. Observations:                  10   AIC:                             61.26
Df Residuals:                       8   BIC:                             61.86
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -5.4519     10.298     -0.529      0.6

D:\Software\Python\lib\site-packages\scipy\stats\_stats_py.py:1736: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


Let $\beta_0 = -5.4519, \beta_1 = 0.9549$. Hence,
$$
\text{expense} = -5.4519 + 0.9549\times\text{income}
$$



$R^2$ is just around $0.672$, that means the model can explain at most $67%$. We might want to add more data / use a different model

In [14]:
-5.4519 + 0.9549 * 40

32.744099999999996